diff --git a/common/urltest/urltest.go b/common/urltest/urltest.go
index 3f4657030..68b620130 100644
--- a/common/urltest/urltest.go
+++ b/common/urltest/urltest.go
@@ -79,7 +79,7 @@ func (s *HistoryStorage) Close() error {
 
 func URLTest(ctx context.Context, link string, detour N.Dialer) (t uint16, err error) {
 	if link == "" {
-		link = "https://www.gstatic.com/generate_204"
+		link = "https://cp.cloudflare.com/generate_204"
 	}
 	linkURL, err := url.Parse(link)
 	if err != nil {
